src/devices/bus/technics/hdae5000.cpp

```diff
@@ -0,0 +1,110 @@
+
+#include "emu.h"
+#include "bus/ata/hdd.h"
+#include "hdae5000.h"
+#include "machine/i8255.h"
+
+namespace {
+
+class hdae5000_device : public device_t, public kn5000_extension_interface
+{
+public:
+	static constexpr feature_type unemulated_features() { return feature::DISK | feature::SOUND; }
+
+	hdae5000_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	virtual void rom_map(address_map &map) override;
+	virtual void io_map(address_map &map) override;
+
+protected:
+	virtual void device_add_mconfig(machine_config &config) override;
+	virtual void device_start() override;
+	virtual void device_reset() override;
+	virtual const tiny_rom_entry *device_rom_region() const override;
+
+private:
+	required_device<ide_hdd_device> m_hdd;
+	required_device<i8255_device> m_ppi;
+	required_memory_region m_rom;
+};
+
+hdae5000_device::hdae5000_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	device_t(mconfig, HDAE5000, tag, owner, clock)
+	, kn5000_extension_interface(mconfig, *this)
+	, m_hdd(*this, "hdd")
+	, m_ppi(*this, "ppi")
+	, m_rom(*this, "rom")
+{
+}
+
+ROM_START(hdae5000)
+	ROM_REGION16_LE(0x80000, "rom" , 0)
+	ROM_DEFAULT_BIOS("v2.01i")
+
+	ROM_SYSTEM_BIOS(0, "v1.10i", "Version 1.10i - July 6th, 1998")
+	ROMX_LOAD("hd-ae5000_v1_10i.ic4", 0x000000, 0x80000, CRC(7461374b) SHA1(6019f3c28b6277730418974dde4dc6893fced00e), ROM_BIOS(0))
+
+	ROM_SYSTEM_BIOS(1, "v1.15i", "Version 1.15i - October 13th, 1998")
+	ROMX_LOAD("hd-ae5000_v1_15i.ic4", 0x000000, 0x80000, CRC(e76d4b9f) SHA1(581fa58e2cd6fe381cfc312c73771d25ff2e662c), ROM_BIOS(1))
+
+	// Version 2.01i is described as having "additions like lyrics display etc."
+	ROM_SYSTEM_BIOS(2, "v2.01i", "Version 2.01i - January 15th, 1999") // installation file indicated "v2.0i" but signature inside the ROM is "v2.01i"
+	ROMX_LOAD("hd-ae5000_v2_01i.ic4", 0x000000, 0x80000, CRC(961e6dcd) SHA1(0160c17baa7b026771872126d8146038a19ef53b), ROM_BIOS(2))
+ROM_END
+
+void hdae5000_device::rom_map(address_map &map)
+{
+	map(0x00000, 0x7ffff).rom().region(m_rom, 0);
+}
+
+void hdae5000_device::io_map(address_map &map)
+{
+	//map(0x130000, 0x13ffff).m(m_hddc, FUNC(?_device::?)); // Hard-drive Controller (model?) IC? on HD-AE5000 board
+	//map(0x160000, 0x16ffff) ... Optional parallel port interface (NEC uPD71055) IC9
+	map(0x160000, 0x160000).lrw8([this](offs_t a) { return m_ppi->read(0); }, "ppi_r0", [this](offs_t a, u8 data) { m_ppi->write(0, data); }, "ppi_w0");
+	map(0x160002, 0x160002).lrw8([this](offs_t a) { return m_ppi->read(1); }, "ppi_r1", [this](offs_t a, u8 data) { m_ppi->write(1, data); }, "ppi_w1");
+	map(0x160004, 0x160004).lrw8([this](offs_t a) { return m_ppi->read(2); }, "ppi_r2", [this](offs_t a, u8 data) { m_ppi->write(2, data); }, "ppi_w2");
+	map(0x160006, 0x160006).lrw8([this](offs_t a) { return m_ppi->read(3); }, "ppi_r3", [this](offs_t a, u8 data) { m_ppi->write(3, data); }, "ppi_w3");
```

Comment:
This would be clearer and less error-prone if you make helpers like:
```c++
	template <unsigned Offset> u8 ppi_r() { return m_ppi->read(Offset); }
	template <unsigned Offset> void ppi_w(u8 data) { m_ppi->write(Offset, data); }
```

Then installed it like:
```c++
	map(0x160000, 0x160000).rw(FUNC(hdae5000_device::ppi_r<0>), FUNC(hdae5000_device::ppi_w<0>));
	map(0x160002, 0x160002).rw(FUNC(hdae5000_device::ppi_r<1>), FUNC(hdae5000_device::ppi_w<1>));
	map(0x160004, 0x160004).rw(FUNC(hdae5000_device::ppi_r<2>), FUNC(hdae5000_device::ppi_w<2>));
	map(0x160006, 0x160006).rw(FUNC(hdae5000_device::ppi_r<3>), FUNC(hdae5000_device::ppi_w<3>));
```

But taking a step back, is this a case of installing an 8-bit device on half the width of a 16-bt bus?  If it is, that’s what `umask` is for:
```c++
	map(0x160000, 0x160007).umask(0x00ff).rw(m_ppi, FUNC(i8255_device::read), FUNC(i8255_device::write));
```